[IMP] survey: allow user to navigate freely

pko-odoo opened the pull request:
This commit allows the user to navigate freely through a
survey with mandatory questions. He will be able to skip
questions and come back to them later. If he still hasn't
answered them at the final submit they will automatically
be displayed with an error message to invite the user to
answer them. The user can navigate through the skipped
questions with a button within the error box.

Task-2605657

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/119315).

thiryr commented on addons/survey/controllers/main.py:
```diff
@@ -322,6 +326,11 @@ def _prepare_survey_data(self, survey_sudo, answer_sudo, **post):
                     'time_limit_minutes': next_page_or_question.time_limit / 60
                 })
 
+            if 'has_skipped_mandatory' in post:
```
Problem with this is that we only set 'has_skipped_mandatory' when submitting the whole form. Meaning the moment you go to the next question, this is forgotten.

The cost of always checking shouldn't be much. But ideally we still need a way to know "This user has already tried submitting once" otherwise everything will be marked the moment you skip it...

If we don't want to store that information, I guess it could be data on the rendered pages but that's a bit dodgy :/

thiryr reviewed (commented):
Heya :wave: 

Seems pretty good. References going in circles with python -> templates -> js -> python doesn't feel ideal for readability... but I don't think there's another way.

I did notice a functional issue though.

thiryr commented on addons/survey/views/survey_templates.xml:
```diff
@@ -336,6 +342,14 @@
+                 class="d-flex align-items-center justify-content-between overflow-hidden border-0 p-3 alert alert-danger">
+                <span t-out="question.constr_error_msg or default_constr_error_msg"/>
+                <button t-if="survey.questions_layout == 'page_per_question'" class="btn btn-sm btn-link"
+                        type="submit" t-att-value="'finish' if len(skipped_questions) == 1 else 'next_skipped'">
```
Just to make sure, we use finish here to stop the timer and that's all? I guess that's fine since they must have already tried to submit in the first place so they probably don't care about the timer for the remaining questions

pko-odoo replied:
finish is used when there is no other question to answer, so here when displaying the last skipped question

thiryr commented on addons/survey/views/survey_templates.xml:
```diff
@@ -212,6 +212,11 @@
+                    <button t-if="skipped_questions" type="submit" class="btn btn-primary"
+                            t-att-value="'finish' if len(skipped_questions) == 1 else 'next_skipped'">
+                        <t t-if="len(skipped_questions) == 1">Submit</t>
+                        <t t-else="">Next Skipped</t>
```
We need to either not display anything in the else case, or hide the original submit button. Currently you get 2 submit buttons in the survey

thiryr commented on addons/survey/views/survey_templates.xml:
```diff
@@ -336,6 +342,14 @@
+                 class="d-flex align-items-center justify-content-between overflow-hidden border-0 p-3 alert alert-danger">
+                <span t-out="question.constr_error_msg or default_constr_error_msg"/>
+                <button t-if="survey.questions_layout == 'page_per_question'" class="btn btn-sm btn-link"
+                        type="submit" t-att-value="'finish' if len(skipped_questions) == 1 else 'next_skipped'">
```
Also should we update the continue button at the bottom of the page too, like we do for sections? And/or maybe add text to specify what the arrow will do.

thiryr commented on addons/survey/controllers/main.py:
```diff
@@ -535,11 +542,23 @@ def survey_submit(self, survey_token, answer_token, **post):
-                    answer_sudo._mark_done()
+                    if survey_sudo.users_can_go_back and answer_sudo.user_input_line_ids.filtered(
+                            lambda a: a.skipped and a.question_id.constr_mandatory):
+                        answer_sudo.write({'last_displayed_page_id': page_or_question_id})
```
Can't we do that at the start of the block to avoid having the same line twice? I don't think it affects any of the computations and it's done in all cases

pko-odoo replied:
Where would you put it exactly ? Actually it's not called when `_mark_done` is and it's not relevant to do so even if it probably doesn't matter

thiryr reviewed (commented):
('This afternoon' is a relative term)
Ok, seems fine to me. Just a couple visual issues and a few nitpicks :)

thiryr reviewed (approved):
Ok, lgtm :)

Thanks for the changes!

flch-odoo commented on addons/survey/static/src/js/survey_form.js:
```diff
@@ -523,6 +529,18 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend(SurveyPreloa
+     * (Those errors are displayed when the user skip mandatory question(s))
+     */
+    _scrollToFirstError: function() {
+        let $firstNonEmptyError = this.$('.o_survey_question_error :not(:empty):first')
+        if ($firstNonEmptyError.length) {
+            let $scrollToElement = $firstNonEmptyError.parents('.js_question-wrapper')
+            this._scrollToError($scrollToElement);
+        }
```
We've decided not to merge new code using jQuery. It should be doable here too.

BTW, I would put it in `ERRORS TOOLS`, near `scrollToError`.

flch-odoo reviewed (commented):
Hello @pko-odoo ,
Thank you for the quick rebase. Here are few suggestions and questions.
Additionally, I did not get the "Next Skipped" button when testing "MyCompany Vendor Certification". Did I miss something?
Also, I find this (historic?) `answer` name for `user_input`s **very** confusing. Do you think we could/should clean that in an additional or preliminary commit as we are here?

pko-odoo commented:
Hello @flch-odoo !

I adapted the code accordingly to your comments :smiley: 
About the Next Skipped button, it's only showed in 'One page per section' that's maybe why you didn't see it ?

flch-odoo reviewed (commented):
Thank you for the update, @pko-odoo !
Here are a few additional ideas, mostly aiming at condensing it a little.
I may come to `prepare_survey_data`and/or `prepare_question_html` later on :disguised_face:...

flch-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -610,6 +611,55 @@ def _get_print_questions(self):
+           and len(self._get_skipped_pages()) == 1
+           and self._get_skipped_pages() == next_page_or_question)):
+            return True
+        return False
```
Now that it is very short, and only used by the controller, shouldn't it be moved there?
Otherwise, it needs at least a docstring imo.

flch-odoo commented on addons/survey/models/survey_question.py:
```diff
@@ -403,11 +403,11 @@ def validate_question(self, answer, comment=None):
+        if not answer and self.question_type not in ['simple_choice', 'multiple_choice']:
+            if self.constr_mandatory and not self.survey_id.users_can_go_back:
+                return {self.id: self.constr_error_msg or _('This question requires an answer.')}
         # because in choices question types, comment can count as answer
```
This comment should be part of the one on top as it the first mention of choices questions is there now.

pko-odoo commented:
Hello @flch-odoo !

I answer here for comments that are now outdated and for which replying is no more possible...

> 1. When I was redirected to a skipped mandatory question, I saw "Submit" in the button, as it was the last one. But as I clicked on it without having answered, I ended up on the next question instead.

Fixed, It will now redisplay the same question until an answer is provided. Currently when on the last question, the right arrow have the same behavior as the submit button. I keep that behavior here.
Personaly, I prefer to keep the same behavior as currently.

> 2. I switched the "MyCompany Vendor Certification" to page_per_question; then skipped the second required question of the first page. When I got back there after submitting, I had no warning on that question. Shouldn't I have had something?

Fixed, I messed up when changing something in the template.

> Now that it is very short (__is_last_skipped_page_or_question_ method), and only used by the controller, shouldn't it be moved there?
> Otherwise, it needs at least a docstring imo.

I don't see why we should put it in the controller? It already has a docstring, should I change it ?

> This function really needs to be updated too: jQuery, var, repeated function calls (will go without jquery)...
> As a quick reminder, we do not want to merge any jQuery outside of tests or if its modifying an existing line in a long function full of jquery, it will be handled in a specific task; but it is not the case here.

I removed the jQuery here but the parameter of the function is still a jQuery class ^^

flch-odoo reviewed (commented):
Hello @pko-odoo ,
Thanks for the update, we're getting there! Here are the answers to the answers.
(perhaps you solved something from these already as I see the "Refresh" button, we'll see :))

pko-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -610,6 +611,46 @@ def _get_print_questions(self):
+        - Avoid displaying a Submit button on the last survey question if they are
+          still skipped questions before.
+        """
+        skipped = self._get_skipped_questions()
+        if self.survey_id.questions_layout == 'page_per_section':
+            skipped = skipped.page_id
+        return len(skipped) == 1 and skipped == next_page_or_question
```
It's now used in two places ^^

flch-odoo replied:
So, we agree it is only used to know whether to show "Submit" or "Continue" on the page?
If so, isn't it a controller's task? (arguing for a controller utility function)

Furthermore, we could make it more robust and return True (meaning "submit") for one_page, or when there are no skipped questions, so you don't have to do the "or" there either: 

```py
... and (not answer_sudo._get_skipped_questions().page_id or
                         answer_sudo._is_last_skipped_page_or_question(page_or_question)
```

flch-odoo reviewed (commented):
Just that one thing left :) Keep it up!

After this round is resolved, as several things did get updated/fixed, shouldn't we ask for a "quick" IN or BE testing pass to be sure?

flch-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -610,6 +611,53 @@ def _get_print_questions(self):
+        return self.user_input_line_ids.filtered(
+            lambda answer: answer.skipped and answer.question_id.constr_mandatory).question_id
+
+    def _is_last_skipped_page_or_question(self, page_or_question):
```
Actually, this has become `is_last_page_or_question` , hasn't it?

pko-odoo replied:
Well no, there could still be a question or more after a skipped question.

flch-odoo replied:
`has_skipped_page_or_question_afterwards`?
@tde-banana-odoo please help us sort this out :) I'd like to call it "_show_submit_button" and move it to the controller ^^  But it is not very different from other methods like this one that are present on the model so...

flch-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -610,6 +611,53 @@ def _get_print_questions(self):
+        """
+        if self.survey_id.questions_layout == 'one_page':
+            return True
+        skipped = self._get_skipped_questions()
+        if self.survey_id.questions_layout == 'page_per_section':
+            skipped = skipped.page_id
+        return not skipped or (len(skipped) == 1 and skipped == page_or_question)
```
Maybe?
```suggestion
        skipped = self._get_skipped_questions()
        if not skipped:
            return True
        if self.survey_id.questions_layout == 'page_per_section':
            skipped = skipped.page_id
        return len(skipped) == 1 and skipped == page_or_question
```

flch-odoo reviewed (commented):
Final note, promise :)
I even ping @tde-banana-odoo for the final step:)

tde-banana-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -610,6 +611,55 @@ def _get_print_questions(self):
+            return True
+        if self.survey_id.questions_layout == 'page_per_section':
+            skipped = skipped.page_id
+        return len(skipped) == 1 and skipped == page_or_question
```
Is the (len) == 1 really necessary ? It is implied by the '==', isn't it ?

tde-banana-odoo reviewed (approved):
LGTM, few linting remarks :)
